documentation/specs/MultithreadedMSBuild/thread-safe-tasks.md

```diff
@@ -0,0 +1,172 @@
+# Thread-Safe Tasks
+
+## Overview
+
+MSBuild's current execution model assumes that tasks have exclusive control over the entire process during execution. This allows tasks to freely modify global process state such as environment variables, the current working directory, and other process-level resources. This design works well for MSBuild's approach of executing builds in separate processes for parallelization. With the introduction of multithreaded execution within a single MSBuild process, multiple tasks can now run concurrently. This requires a new task design to ensure thread safety when multiple tasks access shared process state simultaneously.
+
+To enable this multithreaded execution model, we introduce the `IThreadSafeTask` interface that tasks can implement to declare their thread-safety capabilities. Tasks implementing this interface must avoid using APIs that modify or depend on global process state, as this could cause conflicts when multiple tasks execute concurrently. See [Thread-Safe Tasks API Reference](thread-safe-tasks-api-reference.md) for detailed guidelines. At the same time, the `IThreadSafeTask` interface provides access to the `ExecutionContext` property that allows safe access to global process state. For example, task authors should use `ExecutionContext.GetAbsolutePath(relativePath)` instead of the standard `Path.GetFullPath(relativePath)` to ensure correct path resolution.
+
+## Design
+
+```csharp
+public interface IThreadSafeTask : ITask
+{
+    ITaskExecutionContext ExecutionContext { get; set; }
+}
+```
+
+**Note:** Consider backporting the `IThreadSafeTask` interface to the 17.14 branch to allow a graceful fail when tasks attempt to use it.
+
+The execution context provides essential methods for thread-safe task execution, replacing direct access to global process state:
+
+```csharp
+public interface ITaskExecutionContext
+{
+    AbsolutePath CurrentDirectory { get; set; }
+
+    AbsolutePath GetAbsolutePath(string path);
+    
+    string? GetEnvironmentVariable(string name);
+    IReadOnlyDictionary<string, string> GetEnvironmentVariables();
+    void SetEnvironmentVariable(string name, string? value);
+    
+    Process StartProcess(ProcessStartInfo startInfo);
+    Process StartProcess(string fileName);
+    Process StartProcess(string fileName, string arguments);
```

Comment:
We shouldn't introduce any new Process-related APIs that use a bare string for arguments. It should be `string[]` or ideally `ReadOnlySpan<string>` if at all possible for maximum flexibility.

Reply:
the underlying BCL API accepts `IEnumerable<string>`, so we could just default to that. IMO the main reason to accept `ROS<string>` is to take advantage of the C# compiler's ability to coerce many collection types to that automatically. IEnumerable vs Collection vs ROS<string> isn't _as_ critical to me as making sure we support a collection-based approach to argument-passing.